versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,247 @@
+        PojoCodecProvider.builder().automatic(true).build(),
+        MongoClientSettings.getDefaultCodecRegistry());
+    this.mongoClientSettings = MongoClientSettings.builder()
+      .applyConnectionString(connectionString)
+      .codecRegistry(codecRegistry)
+      .writeConcern(WriteConcern.MAJORITY)
+      .build();
+  }
+
+  /**
+   * Gets a handle to an existing database or get a handle to a MongoDatabase instance if it does not exist. The new
+   * database will be lazily created.
+   * Since MongoDB creates databases and collections if they do not exist, there is no need to validate the presence of
+   * either before they are used. This creates or retrieves collections that map 1:1 to the enumerates in
+   * {@link com.dremio.nessie.versioned.store.ValueType}
+   */
+  @Override
+  public void start() {
+    mongoClient = MongoClients.create(mongoClientSettings);
+    final ClusterType clusterType = mongoClient.getClusterDescription().getClusterSettings().getRequiredClusterType();
+    enableTransactions = (clusterType == ClusterType.REPLICA_SET) || (clusterType == ClusterType.SHARDED);
+    mongoDatabase = mongoClient.getDatabase(databaseName);
+
+    // Initialise collections for each ValueType.
+    collections.put(ValueType.L1, mongoDatabase.getCollection(MongoDbConstants.L1_COLLECTION, L1.class));
+    collections.put(ValueType.L2, mongoDatabase.getCollection(MongoDbConstants.L2_COLLECTION, L2.class));
+    collections.put(ValueType.L3, mongoDatabase.getCollection(MongoDbConstants.L3_COLLECTION, L3.class));
+    collections.put(ValueType.COMMIT_METADATA,
+        mongoDatabase.getCollection(MongoDbConstants.COMMIT_METADATA_COLLECTION, InternalCommitMetadata.class));
+    collections.put(ValueType.KEY_FRAGMENT,
+        mongoDatabase.getCollection(MongoDbConstants.KEY_FRAGMENT_COLLECTION, Fragment.class));
+    collections.put(ValueType.REF, mongoDatabase.getCollection(MongoDbConstants.REF_COLLECTION, InternalRef.class));
+    collections.put(ValueType.VALUE,
+        mongoDatabase.getCollection(MongoDbConstants.VALUE_COLLECTION, InternalValue.class));
+  }
+
+  /**
+   * Closes the connection this manager creates to a database. If the connection is already closed this method has
+   * no effect.
+   */
+  @Override
+  public void close() {
+    if (null != mongoClient) {
+      mongoClient.close();
+    }
+  }
+
+  @Override
+  public void load(LoadStep loadstep) throws ReferenceNotFoundException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public <V> boolean putIfAbsent(ValueType type, V value) {
+    final MongoCollection collection = collections.get(type);
+    if (null == collection) {
+      throw new UnsupportedOperationException(String.format("Unsupported Entity type: %s", type.name()));
+    }
+
+    // TODO: Potentially rework to use updateOne with upsert and $set.
+    try (MongoCursor cursor = collection.find(Filters.eq(Store.KEY_NAME, ((HasId) value).getId())).iterator()) {
```

Comment:
this should be implement correctly so it is a single roundtrip. Let's not add the method until it is implemented correctly.

Reply:
This has been rewritten now.